src/hotspot/share/oops/instanceKlass.cpp

```diff
@@ -2062,6 +2062,57 @@ Method* InstanceKlass::lookup_method_in_all_interfaces(Symbol* name,
   return NULL;
 }
 
+PrintClassClosure::PrintClassClosure(outputStream* st, bool verbose)
+  :_st(st), _verbose(verbose) {
+  ResourceMark rm;
+  _st->print("%-18s  ", "KlassAddr");
+  _st->print("%-4s  ", "Size");
+  _st->print("%-20s  ", "State");
+  _st->print("%-7s  ", "Flags");
+  _st->print("%-12s  ", "LoaderName");
+  _st->print("%-5s  ", "ClassName");
+  _st->cr();
+}
+
+void PrintClassClosure::do_klass(Klass* k)  {
+  ResourceMark rm;
+  // klass pointer
+  _st->print(INTPTR_FORMAT "  ", p2i(k));
+  // klass size
+  _st->print("%-4d  ", k->size());
+  // initialization state
+  if (k->is_instance_klass()) {
+    _st->print("%-20s  ",InstanceKlass::cast(k)->init_state_name());
+  } else {
+    _st->print("%-20s  ","");
+  }
+  // misc flags(Changes should synced with ClassesDCmd::ClassesDCmd help doc)
+  char buf[10];
+  int i = 0;
+  if (k->has_finalizer()) buf[i++] = 'F';
+  if (k->has_final_method()) buf[i++] = 'f';
+  if (k->has_vanilla_constructor()) buf[i++] = 'V';
+  if (k->is_instance_klass()) {
+    InstanceKlass* ik = InstanceKlass::cast(k);
+    if (ik->is_rewritten()) buf[i++] = 'W';
+    if (ik->is_contended()) buf[i++] = 'C';
+    if (ik->has_been_redefined()) buf[i++] = 'R';
+    if (ik->is_shared()) buf[i++] = 'S';
+  }
+  buf[i++] = '\0';
+  _st->print("%-7s  ", buf);
+  // classloader name
+  ClassLoaderData* cld = k->class_loader_data();
+  _st->print("%-12s  ", cld->loader_name());
```

Comment:
For custom class loaders, this will likely print a long class name that will over the 12 character limit, making the output somewhat hard to read.


```
const char* ClassLoaderData::loader_name() const {
   if (_class_loader_klass == NULL) {
     return BOOTSTRAP_LOADER_NAME;
   } else if (_name != NULL) {
     return _name->as_C_string();
   } else {
     return _class_loader_klass->external_name();
   }
}
```

Also, for custom loaders, printing out just the name of the loader class is not sufficient, as multiple loader instances may have the same type.

Maybe we should just remove line 2106? If the user wants to know the class loader, they can use the "-verbose" option of this jcmd.

Reply:
Thanks for review! All comments are addressed.